### PR TITLE
Remove CopyField action from the VxLAN ingress rule.

### PR DESCRIPTION
**Pull request by dpoltavets**

### CODE REVIEW

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        return data;
+    }
+
+    @Override
+    public IPacket deserialize(byte[] data, int offset, int length) {
+        ByteBuffer bb = ByteBuffer.wrap(data, offset, length);
+        // skip 8 bytes
+        bb.getInt();
```
Comment (niksv) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java:63]: incorrect comment, but correct code.
`int` is 4 bytes. not 8

Comment (dpoltavets) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java:63]: Yes. You are right. Comment has been fixed.

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+public class Vxlan extends BasePacket {
+    private static final byte FLAGS = 0x8;
+    private static final int VXLAN_HEADER_SIZE_IN_BYTES = 8;
+    private static final int FIRST_RESERVED_AREA_IN_BYTES = 3;
+    private static final int SECOND_RESERVED_AREA_IN_BITS = 8;
+
+    private int vni;
+
+    @Override
+    public byte[] serialize() {
```
Comment (surabujin) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java:37]: Totally unreadable. You fill only one field and name all other as "reserved"... ugly. Without `VxLan` packet schema somewhere in front of eyes this code is not modifiable.

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java (hunk 1 above)
Comment (surabujin) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java:63]: Why 4?

File: src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/FlowFetcher.java [hunk 3]
```diff
@@ -110,8 +128,10 @@ private void handlePeriodicRequest(Tuple input) throws PipelineException {
             refreshHeap(input, true);
         }
         final CommandContext commandContext = pullContext(input);
-        for (Flow flow : flowsSet) {
-            PingContext pingContext = new PingContext(Kinds.PERIODIC, flow);
+        for (FlowWithTransitEncapsulation flow : flowsSet) {
+            PingContext pingContext = new PingContext(Kinds.PERIODIC, flow.getFlow()).toBuilder()
```
Comment (surabujin) [inline src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/FlowFetcher.java:132]: new ... toBuilder ... build?.. why for?

File: src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/PingProducer.java [hunk 4]
```diff
@@ -66,20 +72,39 @@ private Ping buildPing(PingContext pingContext, FlowDirection direction) {
...
+
+    private int getIslPort(Flow flow, FlowPath flowPath) {
+        List<PathSegment> segments = flowPath.getSegments();
+        if (segments.isEmpty()) {
+            throw new IllegalArgumentException(
+                    format("Path segments not provided, flow_id: %s", flow.getFlowId()));
+        }
+
+        PathSegment ingressSegment = segments.get(0);
+        if (!ingressSegment.getSrcSwitch().getSwitchId().equals(flowPath.getSrcSwitch().getSwitchId())) {
```
Comment (surabujin) [inline src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/PingProducer.java:102]: Why this check is required?

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/service/ping/PingService.java [hunk 5]
```diff
@@ -94,21 +97,40 @@ public IPacket wrapData(Ping ping, byte[] payload) {
...
+            Ethernet ethernet = new Ethernet();
+            ethernet.setPayload(ipv4);
+            ethernet.setEtherType(EthType.IPv4);
+            ethernet.setSourceMACAddress(magicSourceMacAddress);
+            ethernet.setDestinationMACAddress(MacAddress.of(egressSwitch));
+
+            return ethernet;
+        }
+
+        l2.setVlanID(ping.getTransitEncapsulation().getId().shortValue());
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/service/ping/PingService.java]: Fixed.

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import net.floodlightcontroller.packet.BasePacket;
+import net.floodlightcontroller.packet.IPacket;
+
+import java.nio.ByteBuffer;
+
+@lombok.Data
+@EqualsAndHashCode(callSuper = true)
+@NoArgsConstructor
+public class Vxlan extends BasePacket {
+    private static final byte FLAGS = 0x8;
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java:42]: Added description.

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java (hunk 2 above)
Comment (dpoltavets) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/shared/packet/Vxlan.java:50]: Added description.

File: src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/FlowFetcher.java (hunk 3 above)
Comment (dpoltavets) [reply src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/FlowFetcher.java]: Fixed.

File: src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/PingProducer.java (hunk 4 above)
Comment (dpoltavets) [reply src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/PingProducer.java:102]: We verify this when creating requests for installing switch rules. Therefore, I think that there may be cases when this check is needed.

File: src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/FlowFetcher.java [hunk 7]
```diff
@@ -126,17 +148,44 @@ private void handleOnDemandRequest(Tuple input) throws PipelineException {
...
 
-            PingContext pingContext = new PingContext(Kinds.ON_DEMAND, flow).toBuilder()
-                    .timeout(request.getTimeout())
-                    .build();
-            emit(input, pingContext, pullContext(input));
-
+            if (!flow.isOneSwitchFlow()) {
+                Optional<FlowTransitEncapsulation> transitEncapsulation = getTransitEncapsulation(flow);
+                if (transitEncapsulation.isPresent()) {
+                    PingContext pingContext = new PingContext(Kinds.ON_DEMAND, flow).toBuilder()
```
Comment (surabujin) [inline src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/FlowFetcher.java:154]: new ... toBuilder() ... build()... still

Comment (dpoltavets) [reply src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/FlowFetcher.java]: Fixed.

**Pull request merged**